gprofiler/utils/perf.py

```diff
@@ -3,14 +3,82 @@
 # Licensed under the AGPL3 License. See LICENSE.md in the project root for license information.
 #
 
-from gprofiler.exceptions import CalledProcessError
+from enum import Enum
+from pathlib import Path
+from threading import Event
+from typing import List
+
+from gprofiler.exceptions import CalledProcessError, PerfNoSupportedEvent
 from gprofiler.log import get_logger_adapter
 from gprofiler.utils import run_process
-from gprofiler.utils.perf_process import perf_path
+from gprofiler.utils.perf_process import PerfProcess, perf_path
 
 logger = get_logger_adapter(__name__)
 
 
+class SUPPORTED_PERF_EVENTS(Enum):  # pylint: disable=C0103
+    """
+    order here is crucial, the first one we try and succeed - will be used.
+    keep it mind that we should always use `PERF_DEFAULT` as a first try.
+    meaning - keeping with `perf`s' default is always preferred.
+    """
+
+    PERF_DEFAULT = "default"
+    PERF_SW_CPU_CLOCK = "cpu-clock"
+    PERF_SW_TASK_CLOCK = "task-clock"
+
+    def perf_extra_args(self) -> List[str]:
+        if self == SUPPORTED_PERF_EVENTS.PERF_DEFAULT:
+            return []
+        return ["-e", self.value]
+
+
+def perf_default_event_works(tmp_dir: Path, stop_event: Event) -> list:
+    """
+    Validate that `perf record`'s default event actually collects samples.
+
+    We've observed that on some machines the default event `perf record` chooses doesn't actually collect samples.
+    And we generally would not want to change the default event chosen by `perf record`, so before
+    any change we apply to collected sample event, we want to make sure that the default event
+    actually collects samples, and make changes only if it doesn't.
+
+    :param tmp_dir: working directory of this function
+    :return: `perf record` extra arguments to use (e.g. `["-e", "cpu-clock"]`)
+    """
+
+    for event in SUPPORTED_PERF_EVENTS:
+        perf_script_output = None
+
+        try:
+            current_extra_args = event.perf_extra_args() + ["sleep", "0.5"]
```

Comment:
Perhaps add a small doc on the `sleep 0.5` - the reason is to ensure we'll get at least a few samples.

Reply:
86732158d5a0f1cd2c8ec9671d7d9d04e1974a19